### DISCUSSION TITLE
Applying a pressure

### DISCUSSION THREAD

**Discussion by selarem** [idaholab/moose#23528] [Q&A Modules: Solid mechanics]
Hello,
I have applying a pressure as BC as : 

```
[./impact]
    #type = pressure
    #type = FunctionDirichletBC
    type =  FunctionNeumannBC
    variable = disp_y
    boundary = 'impact'
    function = '200*t'
    displacements = 'disp_x disp_y'
  [../]
```

in fact the compuation finishs but without applying anything : everything is zero. The same results obtained with type = pressure.

Of course to check the mesh I have tried FunctionDirichletBC and obtained the expected results.
thanks in advance.

**Comment by jiangwen84** [comment]
I suggest you use ```Pressure``` Action to set up pressure BC. 

If you want to use ```pressure```BC, I remember you need to add ```component``` which should be 1 for your case.

**Reply by selarem** [reply]
like this ?
```
[./impact]
    #type = FunctionDirichletBC
    type =  PressureBC
    variable = disp_y
    component = 1
    boundary = 'impact'
    function = '200*t'
    displacements = 'disp_x disp_y'
  [../]
```
Already tried and gave zero everywhere.

**Reply by selarem** [reply]
```
[BCs]
[./Pressure]
    [./Side1]
      boundary ='impact'
      function = '1.0*t'
      displacements = 'disp_x disp_y'
    [../]
[../]
```
does not work neither.

**Reply by jiangwen84** [reply]
Could you attach your input file here?

I do not know your unit system and time. It could be the function `1.0*t' is too small. I suggest you consider a larger constant value to verify if the pressure is actually applied or not.

**Reply by selarem** [reply]
could not be this, since I have nothing happening (excatly zero everywhere).
The bridge.i example with a pressure applied on the top is working for me !

**Reply by GiudGiud** [reply]
Is there anything holding the system on another side?

**Reply by selarem** [reply]
```
[Mesh]
file = multi6_4.inp
construct_side_list_from_node_list=true
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]



[Modules]
  [./TensorMechanics]
    [./Master]
      [./mech]
        add_variables = true
        strain = SMALL
        additional_generate_output = 'stress_yy stress_xx stress_xy'
      [../]
    [../]
  [../]
[]
[BCs]
[Pressure]
    [Side1]
      boundary ='top'
      function = '1.0*t'
      factor = 5e5
      displacements = 'disp_x disp_y'
    []
[]
 [bottom]
    type = DirichletBC
    variable = disp_x
    boundary = 'right'
    value = 0.0
  []

 [leftx]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value =0.0
   []
[lefty]
    type = DirichletBC
    variable = disp_y
    boundary ='left'
    value =0.0
   []
[]
```

**Reply by jiangwen84** [reply]
What is the initial residual value at the first time step?

**Reply by jiangwen84** [reply]
> What is the initial residual value at the first time step?

Maybe the second time step, because your function is 0 at t=0.

**Reply by GiudGiud** [reply]
@jiangwen84 how does it guess the component here?
```
[BCs]
[Pressure]
    [Side1]
      boundary ='top'
      function = '1.0*t'
      factor = 5e5
      displacements = 'disp_x disp_y'
    []
[]
```

**Reply by selarem** [reply]
> > What is the initial residual value at the first time step?
> 
> Maybe the second time step, because your function is 0 at t=0.

Time Step 0, time = 0

Time Step 1, time = 1e-06, dt = 1e-06

Performing automatic scaling calculation

 0 Nonlinear |R| = 3.030238e-12
 Solve Converged!

Outlier Variable Residual Norms:
  disp_y: 3.030238e-12

Time Step 2, time = 2.5e-06, dt = 1.5e-06
 0 Nonlinear |R| = 7.575595e-12
 Solve Converged!

Outlier Variable Residual Norms:
  disp_y: 7.575595e-12

Time Step 3, time = 4.75e-06, dt = 2.25e-06
 0 Nonlinear |R| = 1.439363e-11
 Solve Converged!

Outlier Variable Residual Norms:
  disp_y: 1.439363e-11

Time Step 4, time = 8.125e-06, dt = 3.375e-06
 0 Nonlinear |R| = 2.462068e-11
 Solve Converged!

**Reply by GiudGiud** [reply]
@selarem add this to Executioner
nl_forced_its = 2
It's a wild guess but maybe the BC isnt properly contributing to the residual on the first nonlinear

**Reply by jiangwen84** [reply]
@GiudGiud is correct. It just means your BC contribution is too small and it actually meet your convergence criterion at 0 nonlinear iteration.   If you converge with the 0 nonlinear, the system is not actually solving anything. 

@selarem Did you remember your phase field fracture input file? Your nl_abs is too large and your system converges at 0 nonlinear. You need to add `nl_forced_its = 2` to enforce the system to be updated.

**Reply by selarem** [reply]
> @selarem add this to Executioner nl_forced_its = 2 It's a wild guess but maybe the BC isnt properly contributing to the residual on the first nonlinear

Not solved with this.

**Reply by jiangwen84** [reply]
@selarem Please send us your full input file and mesh file.

**Reply by jiangwen84** [reply]
> @jiangwen84 how does it guess the component here?
> 
> ```
> [BCs]
> [Pressure]
>     [Side1]
>       boundary ='top'
>       function = '1.0*t'
>       factor = 5e5
>       displacements = 'disp_x disp_y'
>     []
> []
> ```

@GiudGiud The Action should add two kernels, one for each component. For a simple rectangular domain, you would know only one component will be nonzero, so you can just add one there.

**Reply by selarem** [reply]
the problem comes from something related to applying forces. In fact, DirichletBC and FunctionDirichletBC both work fine on the same boundary. However, Pressure and FunctionNeumannBC do not work !

**Reply by jiangwen84** [reply]
@selarem I just cannot agree with you. The MOOSE bridge problem was running without problem, that means the MOOSE pressure BC is implemented correctly. Either something is wrong with your input file, or the way you interpret the results.

**Reply by selarem** [reply]
It is because of the elasticity tensor.
When I consider 

[./elasticity_tensor14]
    type = =ComputeIsotropicElasticityTensor
      youngs_modulus = 3.09
    poissons_ratio = 0.35
  [../]
it works fine. 
But I wanted layers with 'fiber' rotation like:
[./elasticity_tensor25]
    type = ComputeElasticityTensorCP
block = 'Surface2_QUAD4 Surface5_QUAD4'
    #C_ijkl = '11e3 12e3 13e3 22e3 23e3 33e3 44e3 55e3 66e3'
    C_ijkl = '173604000 1423470 1334460 33285600 899760 33340800 3243000 8280000 9384000'
    fill_method = symmetric9
    euler_angle_1 = 0.0
    euler_angle_2 = .0
    euler_angle_3 = 45.0
  [../]

and with this , it never worked.
Strange !

**Reply by jiangwen84** [reply]
I used ComputeElasticityTensorCP in my crystal plasticity simulation before and I was able to apply pressure there without any issue. As you said, it is strange, it should not be related to elasticity tensor. Could you make a simpler problem to test? e.g., one block with moose build-in mesh generator?

**Comment by Rahim-Habibi** [comment]


**Reply by selarem** [reply]
I am not solving for pressure, just want to apply a load.

**Comment by selarem** [comment]
It is all because loose precision !
With a nl_abs_tol =1.0e-8 everything is working fine. I am working with
international units and find that 1.0e-3 is fine for absolute and relative
tolerance. Do not know why I should go so far to have something correct for
a linear elastic computatio.
By the way, what is the integration method used locally (for material) in
nonlinear analysis?
Regards

**Reply by jiangwen84** [reply]
@selarem we already told you about it. The nl_forced_its = 2 was used to force the system to run at least two nonlinear iterations.

I also told you this, with automatic scaling, it is strongly recommended to use nl_rel_tol instead of nl_abs_tol. Depends on your problem, 1e-6 ~ 1e-9 is what I typically use for nl_rel_tol.

Integration typically uses gaussian rule. There is no difference between linear and nonlinear problems.
